cmd/pithosctl/healthz.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright (C) 2018 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	hlz "github.com/gravitational/pithos-app/internal/pithosctl/pkg/healthz"
+
+	"github.com/gravitational/trace"
+	minio "github.com/minio/minio-go"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+const (
+	defaultEndpoint       = "localhost:18080"
+	defaultBucket         = "liveness-check"
+	defaultBucketLocation = "G1"
+	defaultPrefix         = "liveness"
+)
+
+var (
+	healthzConfig hlz.Config
+
+	healthzCmd = &cobra.Command{
+		Use:          "healthz",
+		Short:        "Healthz endpoint for pithos application",
+		SilenceUsage: true,
+		RunE:         healthz,
+	}
+)
+
+// s3Config describes configuration to access an s3 bucket
+type s3Config struct {
+	bucket string
+	client *minio.Client
+}
+
+func init() {
+	pithosctlCmd.AddCommand(healthzCmd)
+	healthzCmd.Flags().StringVar(&healthzConfig.AccessKey, "access-key-id", "", "S3 access key")
+	healthzCmd.Flags().StringVar(&healthzConfig.SecretAccessKey, "secret-access-key", "", "S3 secret key")
+	healthzCmd.Flags().StringVar(&healthzConfig.Endpoint, "endpoint", defaultEndpoint, "S3 endpoint address")
+	healthzCmd.Flags().StringVar(&healthzConfig.Bucket, "bucket", defaultBucket, "S3 bucket name")
+}
+
+func healthz(ccmd *cobra.Command, args []string) error {
+	if err := healthzConfig.Check(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	client, err := initClient(healthzConfig.Endpoint, healthzConfig.AccessKey, healthzConfig.SecretAccessKey)
+	if err != nil {
+		return trace.Wrap(err, "failed to create s3 client")
+	}
+
+	s3Config := &s3Config{
+		bucket: healthzConfig.Bucket,
+		client: client,
+	}
+
+	log.Info("Starting healthz endpoint.")
+
+	http.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
+		if err := livenessProbe(s3Config); err != nil {
+			log.Error(err)
+			w.WriteHeader(http.StatusServiceUnavailable)
+		} else {
+			w.WriteHeader(http.StatusOK)
+		}
+	})
+	return http.ListenAndServe(":8080", nil)
+}
+
+func livenessProbe(s3Config *s3Config) error {
+	// verify that can create S3 bucket
+	if err := s3Config.createBucket(); err != nil {
```

Comment:
Which bucket does this create (if at all)? Do you need to tear it down if `os.Hostname` fails? Can you move it after the `os.Hostname` invocation?

Reply:
`bucketFound, err := s3c.bucketExists()`
I added check for bucket and don't need to tear it down in case of errors. Tear down function will delete bucket with all content there, what is part of health check. Like we want to check CRUD operations over bucket.